libass/ass_blur.c

```diff
@@ -693,142 +429,95 @@ static void coeff_filter(double *coeff, int n, const double kernel[4])
-        mat[ip][jp] = 1;
-        for (int i = 0; i < 4; ++i) {
-            if (i == ip)
+    for (int k = 0; k < n; k++) {
+        double z = 1 / mat[k][k];
+        mat[k][k] = 1;
+        for (int i = 0; i < n; i++) {
+            if (i == k)
                 continue;
 
-            double mul = mat[i][jp] * z;
-            mat[i][jp] = 0;
-            for (int j = 0; j < 4; ++j)
-                mat[i][j] -= mat[ip][j] * mul;
+            double mul = mat[i][k] * z;
+            mat[i][k] = 0;
+            for (int j = 0; j < n; j++)
+                mat[i][j] -= mat[k][j] * mul;
         }
-        for (int j = 0; j < 4; ++j)
-            mat[ip][j] *= z;
+        for (int j = 0; j < n; j++)
+            mat[k][j] *= z;
     }
 }
 
 /**
  * \brief Solve least squares problem for kernel of the main filter
  * \param mu out: output coefficients
- * \param index in: filter tap positions
- * \param prefilter in: supplementary filter type
+ * \param n in: filter kernel radius
  * \param r2 in: desired standard deviation squared
  * \param mul in: scale multiplier
  */
-static void calc_coeff(double mu[4], const int index[4], int prefilter, double r2, double mul)
+static void calc_coeff(double mu[], int n, double r2, double mul)
 {
-    double mul2 = mul * mul, mul3 = mul2 * mul;
+    const double w = 12096;
     double kernel[] = {
-        (5204 + 2520 * mul + 1092 * mul2 + 3280 * mul3) / 12096,
-        (2943 -  210 * mul -  273 * mul2 - 2460 * mul3) / 12096,
-        ( 486 -  924 * mul -  546 * mul2 +  984 * mul3) / 12096,
-        (  17 -  126 * mul +  273 * mul2 -  164 * mul3) / 12096,
+        ((( + 3280 / w) * mul + 1092 / w) * mul + 2520 / w) * mul + 5204 / w,
+        ((( - 2460 / w) * mul -  273 / w) * mul -  210 / w) * mul + 2943 / w,
+        ((( +  984 / w) * mul -  546 / w) * mul -  924 / w) * mul +  486 / w,
+        ((( -  164 / w) * mul +  273 / w) * mul -  126 / w) * mul +   17 / w,
     };
 
-    double mat_freq[14];
-    memcpy(mat_freq, kernel, sizeof(kernel));
-    memset(mat_freq + 4, 0, sizeof(mat_freq) - sizeof(kernel));
-    int n = 6;
-    coeff_filter(mat_freq, n, kernel);
-    for (int k = 0; k < 2 * prefilter; ++k)
-        coeff_blur121(mat_freq, ++n);
-
-    double vec_freq[13];
-    n = index[3] + prefilter + 3;
-    calc_gauss(vec_freq, n, r2);
-    memset(vec_freq + n + 1, 0, sizeof(vec_freq) - (n + 1) * sizeof(vec_freq[0]));
-    n -= 3;
-    coeff_filter(vec_freq, n, kernel);
-    for (int k = 0; k < prefilter; ++k)
-        coeff_blur121(vec_freq, --n);
-
-    double mat[4][4];
-    calc_matrix(mat, mat_freq, index);
-
-    double vec[4];
-    for (int i = 0; i < 4; ++i)
-        vec[i] = mat_freq[0] - mat_freq[index[i]] - vec_freq[0] + vec_freq[index[i]];
-
-    for (int i = 0; i < 4; ++i) {
+    double mat_freq[17] = { kernel[0], kernel[1], kernel[2], kernel[3] };
+    coeff_filter(mat_freq, 7, kernel);
+
+    double vec_freq[12];
+    calc_gauss(vec_freq, n + 4, r2 * mul);
```

Comment:
Worth an `assert(n + 4 <= sizeof(vec_freq)/sizeof(vec_freq[0]))` here.

Reply:
`n + 4` is already copy-paste so using `sizeof(vec_freq)/sizeof(vec_freq[0])` instead of simply `12` don't make much sense. I think I'll go with `assert(n <= 8)` at the start of the function.